test/python/utils/test_deprecation.py

```diff
@@ -229,43 +231,120 @@ def test_deprecate_arg_runtime_warning(self) -> None:
-            del arg1
+        def my_func(
+            arg1: str = "a",
+            arg2: str = "a",
+            arg3: str = "a",
+            new_arg1: str | None = None,
+        ) -> None:
+            del arg2
             del arg3
-            assert new_arg2 == "z"
+            # If the old arg was set, we should set its `new_alias` to that value.
+            if arg1 != "a":
+                self.assertEqual(new_arg1, "z")
+            if new_arg1 is not None:
+                self.assertEqual(new_arg1, "z")
+
+        # No warnings if no deprecated args used.
+        my_func()
+        my_func(new_arg1="z")
 
-        my_func(new_arg2="z")  # No warnings if no deprecated args used.
+        # Warn if argument is specified, regardless of positional vs kwarg.
+        with self.assertWarnsRegex(DeprecationWarning, "arg1"):
+            my_func("z")
         with self.assertWarnsRegex(DeprecationWarning, "arg1"):
-            my_func(arg1="a", new_arg2="z")
+            my_func(arg1="z")
         with self.assertWarnsRegex(DeprecationWarning, "arg2"):
-            # `arg2` should be converted into `new_arg2`.
-            my_func(arg2="z")  # pylint: disable=missing-kwoa
+            my_func("z", "z")
+        with self.assertWarnsRegex(DeprecationWarning, "arg2"):
+            my_func(arg2="z")
+
+        # Error if new_alias specified at the same time as old argument name.
+        with self.assertRaises(TypeError):
+            my_func("a", new_arg1="z")
+        with self.assertRaises(TypeError):
+            my_func(arg1="a", new_arg1="z")
+        with self.assertRaises(TypeError):
+            my_func("a", "a", "a", "z")
 
         # Test the `predicate` functionality.
-        my_func(new_arg2="z", arg3="okay value")
+        my_func(arg3="okay value")
         with self.assertWarnsRegex(DeprecationWarning, "arg3"):
-            my_func(new_arg2="z", arg3="deprecated value")
+            my_func(arg3="deprecated value")
+        with self.assertWarnsRegex(DeprecationWarning, "arg3"):
+            my_func("z", "z", "deprecated value")
+
+    def test_deprecate_arg_runtime_warning_variadic_args(self) -> None:
+        """Test that `@deprecate_arg` errors when defined on a function using `*args` but can
+        handle `**kwargs`.
+
+        We know how to support *args robustly via `inspect.signature().bind()`, but it has worse
+        performance (see the commit history of PR #9884). Given how infrequent we expect
+        deprecations to involve *args, we simply error.
+        """
+
+        with self.assertRaises(ValueError):
+
+            @deprecate_arg("my_kwarg", since="9.99")
+            def my_func1(*args: int, my_kwarg: int | None = None) -> None:
+                del args
+                del my_kwarg
+
+        @deprecate_arg("my_kwarg", since="9.99")
+        def my_func2(my_kwarg: int | None = None, **kwargs) -> None:
+            c = 0
```

Comment:
```suggestion
            c = 0  # pylint: disable=unused-variable
```
This is causing a lint failure in https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=45887&view=logs&j=86d13450-b8b4-5d9a-22f1-9f7e71062a71&t=a7f62dd9-9d78-5db3-3068-aed07887bd76&l=14 . Can you also add a comment explaining why it's here (seems like from https://github.com/Qiskit/qiskit-terra/pull/9884#discussion_r1166099260 ?).

Reply:
Good catch. Thanks!